extensions/sql-database-projects/src/tasks/sqlDatabaseProjectTaskProvider.ts

```diff
@@ -0,0 +1,163 @@
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import * as path from 'path';
+import * as constants from '../common/constants';
+
+/**
+ * Extends to vscode.TaskDefinition to add the filePath and fileDisplayName properties.
+ * This is used to identify the task and provide the file path and display name for the task.
+ */
+interface SqlprojTaskDefinition extends vscode.TaskDefinition {
+	fileDisplayName: string;
+	runCodeAnalysis?: boolean;
+}
+
+/**
+ * This class implements the vscode.TaskProvider interface to provide tasks for SQL database projects.
+ * It creates tasks for building SQL database projects and running code analysis on them.
+ */
+export class SqlDatabaseProjectTaskProvider implements vscode.TaskProvider {
+	private watchers: vscode.FileSystemWatcher[] = [];
+	private sqlTasks: Thenable<vscode.Task[]> | undefined = undefined;
+
+	/**
+	 * This method is used to create a file system watcher for the .sqlproj files in the workspace.
+	 * It is used to watch for changes to the .sqlproj files and update the tasks accordingly.
+	 * @param workspaceRoots The workspace roots to watch for .sqlproj files
+	 * @returns A file system watcher for the .sqlproj files
+	 * @throws Error if the workspace root is not a valid path
+	 * @throws Error if the file system watcher cannot be created
+	 * @throws Error if the file system watcher is not supported
+	 * */
+	constructor(workspaceRoots: readonly vscode.WorkspaceFolder[] | undefined) {
+		if (!workspaceRoots) {
+			return;
+		}
+
+		for (const root of workspaceRoots) {
+			const pattern = path.join(root.uri.fsPath, '**', '*.sqlproj');
+			const watcher = vscode.workspace.createFileSystemWatcher(pattern);
+			watcher.onDidChange(() => this.sqlTasks = undefined);
+			watcher.onDidCreate(() => this.sqlTasks = undefined);
+			watcher.onDidDelete(() => this.sqlTasks = undefined);
+			this.watchers.push(watcher);
+		}
+	}
+
+	/**
+	 * This method is used to dispose of the file system watcher.
+	 * It is called when the task provider is disposed.
+	 */
+	public dispose() {
+		this.watchers?.forEach(watcher => watcher.dispose());
+	}
+
+	/**
+	 * This method is called when the task provider is registered.
+	 * It is used to provide the tasks for the provider.
+	 * @returns The task type for this provider
+	 */
+	public provideTasks(): Thenable<vscode.Task[]> | undefined {
+		if (!this.sqlTasks) {
+			this.sqlTasks = this.createTasks();
+		}
+		return this.sqlTasks;
+	}
+
+	/*
+	 * This method is called when the task is resolved.
+	 * It is used to resolve the task and return the task object.
+	 */
+	public resolveTask(task: vscode.Task): vscode.Task | undefined {
+		if (task.definition.type === constants.sqlProjTaskType) {
+			const definition: SqlprojTaskDefinition = <any>task.definition
+			if (!definition.filePath || !definition.fileDisplayName) {
```

Comment:
If we do the opposite we can avoid a `return undefined;` (and rely on the one on line 83)

```
if (definition.filePath && definition.fileDisplayName) {
    return this.getTask(definition);
}